### DISCUSSION TITLE
Different Characters from Old Password Policy

### DISCUSSION THREAD

**Discussion by lrodriguez14** [keycloak/keycloak#13925] [Ideas]
Hello,

I think that a new password policy can be added to Keycloak to make the system more secure. Since we are already storing the old password for the recently use policy. We can define a new policy called like **"Old Password Different Characters"** where we can define a number of restricting the number of characters from the old password.

For example we set the policy and set the number 2, so if the user had the password **Puppy1** and sets the new as **Puppy2** the system will not let him since just one character changed from the old one.

I think that this policy could be really helpful adding more security to the system.

Regards,
-Leonardo R.

**Comment by darius-m** [comment]
That sounds like a *really* bad idea. Passwords *MUST* never be stored in plain text format, and salted hashes are used instead (i.e. concatenate a randomly generated string to the user's password, use a password hashing algorithm such as `bcrypt` on the resulting string, and then store the salt and the password hash in the database). This means that you cannot have the old password to compare the new password to. The old password policies rely on the fact that using the same hashing algorithm and salt you will always get the same hash for the same password, but changing even one letter will result in a completely different hash.

This is because password databases are unfortunately commonly enough leaked, so having strong hashing algorithms and salting is required to slow an attacker's ability to retrieve the original passwords. Usually, password policies should require longer passwords (as in more characters) that contain special characters to increase the number of possibilities that must be checked in order to retrieve the original password using brute force. As a rule of thumb (assuming a good hashing algorithm is used), the complexity of finding a password using a brute force attack is usually on the magnitude of `set_size ^ length` attempts. For passphrases that only contain regular words, the `set_size` is given by the number of words in a dictionary, and `length` is the number of words in the passphrase; for character passwords, the `set_size` is the number of characters in the character set used (for English that would be 26 if using only lowercase or uppercase letters, 52 when using both lowercase and uppercase, 62 if numbers are added, and so on), and the `length` is the number of characters in the password.

It might make more sense for you to train people to use passphrases (4-5 words, can be from different languages) that incorporate some special characters inside them to make using dictionaries harder or even impossible (there is a well known [xkcd comic](https://xkcd.com/936/) on password strength and someone's ability to remember their password).

Even better, they could use a password manager to store passwords that are extremely long and hard to break, without having to remember them.
